### ISSUE TITLE
WebGL2: panicked at 'Tex storage 2D multisample is not supported'

### ISSUE THREAD

**Issue by superdump** [gfx-rs/wgpu#2149]
**Description**
When using wgpu master with WebGL2 and MSAA I'm getting: `panicked at 'Tex storage 2D multisample is not supported', C:\Users\super\.cargo\registry\src\github.com-1ecc6299db9ec823\glow-0.11.0\src\web_sys.rs:2168:9`

**Repro steps**

See the attached zip of a wasm WebGL2 demo. Unpack, serve, browse, see panic.

**Expected vs observed behavior**
Expected: no panic. Observed: panic in both Chrome and Firefox. :)

**Extra materials**
[webgl2-multisample-panic.zip](https://github.com/gfx-rs/wgpu/files/7461193/webgl2-multisample-panic.zip)

**Platform**
wgpu master @ ca7ca749cdb2b5d88512be2b1ec4e44f5a88dfe0, WebGL2, Windows 10

**Comment by mrk-its** [comment]
I have the same for `texture_pipelined` bevy example, it doesn't panic after disabling msaa

**Comment by kvark** [comment]
Problem is very similar to #2137. We want to enable COPY_DST usage for all textures because that's how we zero-initialize them. But on WebGL specifically, we can't create a multi-sampled texture with COPY_DST and COLOR_TARGET.

The solution I tentatively propose is to split the texture initialization into 2 paths:
  1. the current path, which initializes via copies
  2. for MSAA textures, surface textures, and depth/stencil - initialize via render pass clears. I believe this is also faster on desktop, and it doesn't require any new usages, thus has WebGL2 compatibility.

cc @Wumpf

**Comment by superdump** [comment]
In case it impacts any prioritisation, I just wanted to note  that bevy will need this for WebGL2 support for its upcoming 0.6 release, so there is interest and significant value for bevy there. No expectations of course, and if I could figure out all the needed changes and how to choose one or the other path then I could probably implement it myself.

**Comment by kvark** [comment]
There is a quick thing that we could do, and a real fix that we could do. Both much depend on @Wumpf guidance.
The quick thing is having some option to disable zero-initialization completely. This will be unsafe and native-only. Could be exposed as a device feature. @superdump you could try to prototype this, if you feel motivated.
The real fix is splitting the texture initialization paths in 2, like I described above: via copies and via render target clears.

Finally, Bevy has an option to not do MSAA at start of 0.6, at least not on WebGL target.

**Comment by Wumpf** [comment]
started to take a stab at this, don't know how long it will take and if I can do something partial to unblock the particular issue
We should add a zero-init disable extension regardless I think. At least right now the impl isn't exactly geared towards performance so it can cause issues regardless.

**Comment by superdump** [comment]
Is the zero initialisation stuff a WebGPU requirement or a debug nice to have? If not the former then shouldn’t it be up to users to initialise buffers or not? If the former then maybe it’s still desirable for native anyway to have control over initialisation as well as having an easy way to enable it for debugging. :)

**Comment by Wumpf** [comment]
it is a WebGPU requirement. And a surprisingly nasty one 😄

**Comment by maxammann** [comment]
I just upgarded to 0.12 and encountered this issue. As far as I understand this block multisampling when using WebGL. This is not a huge blocker for me as WebGL is not my primary target. For the time being I can just disable multisampling on WebGL.

**Comment by superdump** [comment]
@kvark I see this was fixed - fantastic! What's the plan for the next release? Any chance of a 'quick' 0.12.1 with this in it?

**Comment by kvark** [comment]
I'll get something rolling...

**Comment by kvark** [comment]
Actually, there is a bit of a problem. #2307 is breaking API (feature name changed, some errors as well).
@Wumpf how feasible would it be to port it to 0.12 in a non-breaking way?

**Comment by maxammann** [comment]
The fix works :tada:

**Comment by Wumpf** [comment]
should be possible, feature name and errors are arguably rather cosmetic and can be reverted. Would be mostly worried about introducing new bugs to an old release

**Comment by Wumpf** [comment]
I think that would do #2327

**Comment by gilescope** [comment]
Alas having studied the existing examples and the PR it's not clear how to get around this error (I'm on latest release v14). Would be great to have a webgl example with texture aliasing as well as just the line one. Is there any project out there I can look at to see what I should be doing?

**Comment by maxammann** [comment]
You should not get this error when enabling multisampling. I'm using multisampling with webgl successfully here: https://github.com/maplibre/maplibre-rs
